File: linkedlist/stack.js
var Stack = function() {
    var Node = function(val) {
        this.val = val;
        this.next = null;
    };

    var head = null;

    this.push = function(val) {
        var node = new Node(val);
        node.next = head;
        head = node;
    };

    this.pop = function() {
        if (!head) {
            return;
        }
        var val = head.val;
        head = head.next;
        return val;
    };

    this.traverse = function(fn) {
        var cur = head;
        while (cur) {
            fn(cur.val);
            cur = cur.next;
        }
    };

    this.empty = function(fn) {
        while (head) {
            fn(head.val);
            head = head.next;
        }
    };
};

assert = require('assert');

var log;
var visitor = function(val) {
    log.push(val);
};

var stack = new Stack;
assert.equal(stack.pop(), undefined);

stack.push(1);
stack.push({foo: 'bar'});

log = [];
stack.traverse(visitor);
assert.deepEqual(log, [{foo: 'bar'}, 1]);

assert.deepEqual(stack.pop(), {foo: 'bar'});
assert.equal(stack.pop(), 1);

stack.push(1);
stack.push({foo: 'bar'});

log = [];
stack.empty(visitor);
assert.deepEqual(log, [{foo: 'bar'}, 1]);
assert.equal(stack.pop(), undefined);

File: linkedlist/stack.py
class Stack(object):
    class Node(object):
        def __init__(self, val):
            self.val = val

    def __init__(self):
        self.head = None

    def push(self, val):
        node = self.Node(val)
        node.next = self.head
        self.head = node

    def pop(self):
        if not self.head:
            return None
        val = self.head.val
        self.head = self.head.next
        return val

    def traverse(self, fn):
        cur = self.head
        while cur is not None:
            fn(cur.val)
            cur = cur.next

    def empty(self, fn):
        while self.head is not None:
            fn(self.head.val)
            self.head = self.head.next

if __name__ == '__main__':

    log = None
    def visitor(val):
        log.append(val)

    stack = Stack()
    assert stack.pop() is None

    stack.push(1)
    stack.push(dict(foo='bar'))

    log = []
    stack.traverse(visitor)
    assert log == [{'foo': 'bar'}, 1]

    assert stack.pop() == dict(foo='bar')
    assert stack.pop() == 1

    stack.push(1)
    stack.push(dict(foo='bar'))


    log = []
    stack.empty(visitor)
    assert log == [{'foo': 'bar'}, 1]
    assert stack.pop() is None
